### ISSUE TITLE
v.rectify (Processing: GRASS provider) consistently fails with "Required parameter <output> not set"

### ISSUE THREAD

**Issue by benducke** [qgis/QGIS#45149]
### What is the bug or the crash?

The GRASS module v.rectify is not correctly wrapped/parsed/executed by the GRASS provider in Processing.

### Steps to reproduce the issue

1. Open Processing Toolbox.
2. Launch GRASS' "v.rectify"
3. Fill in minimal options: input layer and either a text file with four source/target points or the points themselves in the text field.
4. Run "v.rectify"

The tool will always quit with an error. Take a look at the GRASS console output:
The command line for "v.rectify" passed to GRASS GIS is missing the essential "output=" option, which causes the tool to throw an error: "Required parameter <output> not set".

I use these coordinates (plus "active flag") to transform from a local system with small coordinate ranges to UTM 37N (you can try that with any layer, although it would produce wrong results; the point here is that I can currently produce no result at all):

-637.240 580.860 432843.854 3823560.989 1
-962.079 1086.419 432522.788 3824068.656 1
-1369.031 837.071 432114.226 3823822.337 1
-1877.673 775.273 431605.364 3823764.216 1

### Versions

Tested with: QGIS 3.16 and 3.20, Windows 10 and Arch Linux (AMD64).

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

I have tried both using "Inline control points" and reading points from a file (alternate options).
I have also checked the associated "description" file ("v.rectify.txt") used by the GRASS provider in Processing: It is fine, but misses the "group=" option (see below).

There are two things that make me suspicious:
1. GRASS' "v.rectify" has an additional "imagery type" input option ("group="), which is _not_ wrapped by Processing. Perhaps this messes up the parsing sequence of GRASS module options?
2. "v.rectify" has a weird "inline" option of type "string", which is used to enter control points directly into the UI. This _is wrapped correctly_ by Processing (and present in "v.rectify.txt"), and it is represented properly in the Processing UI for the tool. But I wonder if it throws of the parser that assembles the GRASS command line? I do notice that the "inline points" are _not present_ on the parsed GRASS command line, as produced by the Processing provider.

Currently, "v.rectify" is the only tool in QGIS that allows transformation of an existing vector layer by control points. Therefore, it is a pity that it does not run properly.

**Comment by nicogodet** [comment]
Could you try this ?https://github.com/qgis/QGIS/blob/2da03deeec921514b42119721eba97891d30b556/python/plugins/grassprovider/ext/v_rectify.py#L38-L51

```diff
-     alg.processCommand(parameters, context, feedback, True)
+     alg.processCommand(parameters, context, feedback)
```
Last arg is `delOutputs`

It fails because the bellow condition is `False` and outputs are not defined.
https://github.com/qgis/QGIS/blob/2da03deeec921514b42119721eba97891d30b556/python/plugins/grassprovider/Grass7Algorithm.py#L647

**Comment by benducke** [comment]
Good catch! The above modification allows me to run "v.rectify" successfully, if (and only if) I use option "points=" to pass the control points via a text file. However, typing the point list into the tool's UI (using the "inline_points" option) does _not_ work (it produces an error: "ERROR: Please select a group or give an input file." from "v.rectify"). This is strange, because "v_rectify.py" does contain the code to handle this case: It writes the contents of "inline_points" into a temporary text file ("tempPoints"), and then it adds the path to the temp file to "parameters['points']". For some reason, though, this does not result in "points=" being actually added to the GRASS command line at runtime.

**Comment by nicogodet** [comment]
@benducke Could you provide:

- sample data,
- expected results,
- step to reproduce,
- processing parameters ?

**Comment by benducke** [comment]
Sure. All of those points are identical to my very first message in this ticket, though, with one exception:

3. Fill in minimal options: input layer and the points themselves in the text field "Inline control points":

-637.240 580.860 432843.854 3823560.989 1
-962.079 1086.419 432522.788 3824068.656 1
-1369.031 837.071 432114.226 3823822.337 1
-1877.673 775.273 431605.364 3823764.216 1

The expected result would be a transformed new version of the input layer.
The actual result is an error message "ERROR: Please select a group or give an input file."

The reason for the error is that the string type option "inline_points" (defined in "v.rectify.txt") is not properly forwarded to the GRASS command line.

Let me know if this is too convoluted. Then we can just close this ticket, and I will file a new one to focus on this issue.
Many thanks for your help!

**Comment by nicogodet** [comment]
I need an input layer too.

But try this too:
https://github.com/qgis/QGIS/blob/2da03deeec921514b42119721eba97891d30b556/python/plugins/grassprovider/ext/v_rectify.py#L38-L51

```diff
- parameters['points'] = tempPoints
+ parameters['points'] = pointsName
```

So the complete function should look like:
```python
def processCommand(alg, parameters, context, feedback):
    # handle inline points
    inlinePoints = alg.parameterAsString(parameters, 'inline_points', context)
    if inlinePoints:
        # Creates a temporary txt file
        pointsName = getTempFilename()

        # Inject rules into temporary txt file
        with open(pointsName, "w") as tempPoints:
            tempPoints.write(inlinePoints)
        alg.removeParameter('inline_points')
        parameters['points'] = pointsName

    alg.processCommand(parameters, context, feedback)
```

**Comment by benducke** [comment]
That fixes the problem. I have tested running "v.rectify" through its Python wrapper using a number of different layers and combinations of options, and AFAICT, it runs exactly like it should now. The ticket can be closed. I would love to see these fixes in the next revisions of QGIS 3.16 and 3.20. Great work!

**Comment by nicogodet** [comment]
Could you provide a dataset (input layer, inline control points, parameters, output) so that I can add a test ?

**Comment by nicogodet** [comment]
Not stale

**Comment by nicogodet** [comment]
#45517
